daffodil-cli/src/main/scala/org/apache/daffodil/Main.scala

```diff
@@ -929,7 +930,7 @@ object Main {
 
             processor = processor.withValidationMode(validate)
 
-            setupDebugOrTrace(processor.asInstanceOf[DataProcessor], conf)
+            val debugProcessor = setupDebugOrTrace(processor.asInstanceOf[DataProcessor], conf)
```

Comment:
In this scope, processor is a mutable var so you may not need `debugProcessor`.  Perhaps line 933 can replace processor exactly like line 931 already replaces processor:

`processor = setupDebugOrTrace(processor.asInstanceOf[DataProcessor], conf)`

That change should work unless the compiler complains about types, and it also makes me think the method name might look better if it was renamed to `withDebugOrTrace`.

There's a region further down where processor is a val, so debugProcessor is really needed there.  Mike, Steve, would you prefer that both regions use a mutable var processor or an immutable val processor and a val debugProcessor?  I think it's best practice to use val and careful variable naming to make it clear a value is changing, but var is probably fine in a small lexical scope especially when processor is replaced 2 times, not just 1 time.

Reply:
Yeah, the var approach seems reasonable to me in this case. Especially since `debugProcessor` might not even have debugging enabled depending on the provide conf options, so it could potentially be confusing. It's more like `maybeDebugProcessor` which is ugly and still confusing.

An alternative, though maybe a bit more invasive, might be to do what we do with the performance command, which looks like this:

```scala
        val processor = {
          if (performanceOpts.parser.isDefined) {
            createProcessorFromParser(performanceOpts.parser(), performanceOpts.path.toOption, validate)
          } else {
            val tunables = DaffodilTunables.configPlusMoreTunablesMap(performanceOpts.tunables, optDafConfig)
            createProcessorFromSchema(performanceOpts.schema(), performanceOpts.rootNS.toOption, performanceOpts.path.toOption, tunables, validate)
          }
        }.map{ _.withExternalVariables(combineExternalVariables(performanceOpts.vars, optDafConfig)) }
         .map{ _.withValidationMode(validate) }
```
We could just add a `.map { withDebugOrTrace(_, conf) }` to the end of that.

This way all processor setup happens in one spot for each subcommand, the processor variable can always be a val, and is consistent among the different subcommands.

Reply:
I will add the `.map { withDebugOrTrace(_, conf) }`. But with that should I also remove the line that does `var processor = proc` and just use `proc` instead?

Reply:
Makes sense, that would use the same patern as the unparse and performance subcommands

Reply:
So `processor` would now need to be `Option[DataProcessor]` instead of `Option[DFDL.DataProcessor]` right?

Reply:
So I was able to get it work like this:

```scala
val processor: Option[DataProcessor] = {
  if (parseOpts.parser.isDefined) {
    createProcessorFromParser(parseOpts.parser(), parseOpts.path.toOption, validate)
  } else {
     val tunables = DaffodilTunables.configPlusMoreTunablesMap(parseOpts.tunables, optDafConfig)
      createProcessorFromSchema(parseOpts.schema(), parseOpts.rootNS.toOption, parseOpts.path.toOption, tunables, validate)
  }
}.map{ _.withExternalVariables(combineExternalVariables(parseOpts.vars, optDafConfig)) }
 .map{ _.withValidationMode(validate) }
 .map{ proc => withDebugOrTrace(proc.asInstanceOf[DataProcessor], conf) }
```

Doing `.map{ withDebugOrTrace(_.asInstanceOf[DataProcessor], conf) }` or `.map{ withDebugOrTrace(_, conf) }` caused errors. Is this what would be desired or would you want me to add `withDebugger` defintions to `DataProcessorBase` trait that you just mentioned?

Reply:
I have no strong preference. It might be cleaner if the withDebugger defs were added to the trait though. I'm not sure if there was a reason for those to not be added originally. Might have just been an oversight.

Reply:
So when I add them in I still need to do the `.asInstanceOf[DataProcessor]`, is there something I was missing because up until `withDebugOrTrace` I believe that processor is treated as a `DFDL.DataProcessor` right? So that would always be needed unless converting other methods to switch the type of `DataProcessor` being set right?

Reply:
You might need to change `withDebugOrTrace` to accept a `DFDL.DataProcessor` instead of a `DataProcessor`.

Reply:
Oh okay I think I got will push that up once I get it fully tested locally to make sure it works fully.

Reply:
@tuxji @stevedlawrence The rename of `setupDebugOrTrace` to `withDebugOrTrace`, adding `withDebugger` method to`DataProcessorBase` and updating the way the processor is set is up if you want to look it over.